cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/NotificationSubscriberService.java

```diff
@@ -0,0 +1,341 @@
+      this.messagingContext = new MultiThreadMessagingContext(messagingService);
+      emptyFetchCount = 0;
+      failureCount = 0;
+    }
+
+    @Override
+    public void run() {
+      while (isRunning()) {
+        try {
+          long sleepTime = fetchNewNotifications();
+          // Don't sleep if sleepTime returned is 0
+          if (sleepTime > 0) {
+            TimeUnit.MILLISECONDS.sleep(sleepTime);
+          }
+        } catch (InterruptedException e) {
+          // It's triggered by stop
+          Thread.currentThread().interrupt();
+        }
+      }
+    }
+
+    /**
+     * Fetch new notifications and update job queue
+     *
+     * @return sleep time in milliseconds before next fetch
+     */
+    private long fetchNewNotifications() {
+      try {
+        MessageFetcher fetcher = messagingContext.getMessageFetcher();
+        emptyFetchCount++;
+        try (CloseableIterator<Message> iterator = fetcher.fetch(NamespaceId.DEFAULT.getNamespace(),
+                                                                 topic, 100, messageId)) {
+          LOG.trace("Fetch with messageId = {}", messageId);
+          String currentMessageId = null;
+          try {
+            while (iterator.hasNext() && isRunning()) {
+              emptyFetchCount = 0;
+              Message message = iterator.next();
+              Notification notification;
+              try {
+                notification = GSON.fromJson(new String(message.getPayload(), StandardCharsets.UTF_8),
+                                             Notification.class);
+                processNotification(notification);
+              } catch (JsonSyntaxException e) {
+                LOG.warn("Failed to decode message with id {}. Skipped. ", message.getId(), e);
+              }
+              // Record current message's Id no matter decode is successful or not,
+              // so that this message will be skipped in next fetch
+              currentMessageId = message.getId();
+            }
+          } catch (Exception e) {
+            LOG.warn("Failed to fetch new notification. Will retry in next run", e);
+            failureCount++;
+            // Exponential strategy doesn't use the time component, so doesn't matter what we passed in as startTime
+            return scheduleStrategy.nextRetry(failureCount, 0);
+          }
+          if (currentMessageId != null) {
+            // Update messageId with the last message's Id
+            messageId = currentMessageId;
+          }
+        }
+      } catch (Exception e) {
+        LOG.warn("Failed to get notification. Will retry in next run", e);
+        failureCount++;
+        // Exponential strategy doesn't use the time component, so doesn't matter what we passed in as startTime
+        return scheduleStrategy.nextRetry(failureCount, 0);
+      }
+      failureCount = 0;
+
+      if (runFlag.compareAndSet(false, true)) {
```

Comment:
Why need this flag? You can pretty much just launch the program. Launching program is thread safe.